vllm/v1/engine/parallel_sampling.py

```diff
@@ -0,0 +1,370 @@
+                    # cleanup parent request bookkeeping.
+                    agg_outputs.append(out)
+                    del self.parent_reqs[req_id]
+                # Cleanup child request bookkeeping.
+                del self.child_reqs[c_req_id]
+            else:
+                # Not a parallel sampling request output
+                agg_outputs.append(c_out)
+        return agg_outputs
+
+    def get_num_unfinished_requests(self, num_core_reqs: int) -> int:
+        """Get the number of unfinished requests, correcting for parallel
+           sampling.
+        
+        Args:
+          num_core_reqs: The number of unfinished requests in the engine core.
+        
+        Returns:
+          Number of unfinished requests, where each parallel sampling req 
+          counts as 1
+        """
+        return num_core_reqs + len(self.parent_reqs) - len(self.child_reqs)
+
+
+async def generate_parallel_sampling_async(
+    generate: AsyncGenerateMethodType,
+    prompt: PromptType,
+    sampling_params: SamplingParams,
+    request_id: str,
+    lora_request: Optional[LoRARequest] = None,
+    trace_headers: Optional[Mapping[str, str]] = None,
+    prompt_adapter_request: Optional[PromptAdapterRequest] = None,
+    priority: int = 0,
+) -> AsyncGenerator[RequestOutput, None]:
+    """Generate completions for async parallel sampling requests."""
+    parent_req = ParallelSamplingRequest(request_id, sampling_params)
+
+    # Aggregate generators for n child requests
+    gens: List[AsyncGenerator[RequestOutput, None]] = []
+    for idx in range(parent_req.n):
+        c_req_id, c_params = parent_req.get_child_info(idx)
+        child_gen = generate(
+            prompt=prompt,
+            sampling_params=c_params,
+            request_id=c_req_id,
+            lora_request=lora_request,
+            trace_headers=trace_headers,
+            prompt_adapter_request=prompt_adapter_request,
+            priority=priority,
+        )  # type: ignore
+        gen = parent_req.wrap_child_async_generator(child_gen, idx)
+        gens.append(gen)
+
+    # Merge generators
+    async for _, out in merge_async_iterators(*gens):
+        yield out
+
+
+def add_request_parallel_sampling(
+    add_request: SyncAddRequestMethodType,
+    parallel_mgr: SyncParallelSamplingManager,
+    request_id: str,
+    prompt: PromptType,
+    params: Union[SamplingParams, PoolingParams],
+    arrival_time: Optional[float] = None,
+    lora_request: Optional[LoRARequest] = None,
+    trace_headers: Optional[Mapping[str, str]] = None,
+    prompt_adapter_request: Optional[PromptAdapterRequest] = None,
+    priority: int = 0,
+) -> None:
+    """Add sync parallel sampling request."""
+    req = ParallelSamplingRequest(request_id, params)
+    parallel_mgr.register_parent_request(req)
+    # Add n child requests with unique request IDs & random seeds and n=1
+    for idx in range(req.n):
+        c_req_id, c_params = req.get_child_info(idx)
```

Comment:
use full names? the `c_` is less readable to me for some reason (and elsewhere)
```suggestion
        child_req_id, child_params = req.get_child_info(idx)
```

Reply:
Agreed, fixed manually